revamp 'Hello world' in Agda

pnlph opened the pull request:
This PR extends the documentation for the 'Hello World' according to issue #4556.

pnlph committed:
revamp 'Hello world' in Agda

This PR extends the documentation for the 'Hello World' according to issue #4556.

pnlph committed:
add Agdapad to installation page

pnlph committed:
add warning ieee754 haskell library

pnlph committed:
remark compatible version stdlib

pnlph committed:
add where to go from here

pnlph committed:
add cabal install ieee754

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+
+  open import Agda.Builtin.Nat using (Nat ; zero ; suc)
+
+  data Vec (A : Set) : ℕ → Set where
```
Please make sure the code here actually type-checks ;)

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+
+This code has 6 declarations:
+
+1. top level module declaration
```
It wouldn't hurt to point it out specifically, ie. say that "module hello-world-dep where" is the module header.

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+============================
+
+We will model the notion of vector space with Agda. Roughly speaking, a vector
+is a list of objects with a determined length.
```
I'm not sure if this is good as a first example. It is certainly good as an example of what's *unique* about Agda, but for someone who is just getting started it makes more sense to first have an example of something you can also do in other languages, in order to get used to the syntax of Agda.

pnlph replied:
I can add a previous simpler example, which one do you have in mind?

andreasabel replied:
"the notion of vector space"  Actually, `Vec` does not implement the notion of vector space, which is an algebraic structure. https://en.wikipedia.org/wiki/Vector_space#Definition

The classic "hello world of dependent types" is `Vec` with a total lookup function.
```agda
lookup : Vec A n -> Fin n -> A
lookup (a :: as) zero = a
lookup (a :: as) (suc i) = lookup as i
```
You can do this in other programming languages, but it won't be a total function then.

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+The constructors ``[]`` and ``_∷_``
+-----------------------------------
+
+Constructors are declared in new lines and indented with two spaces.
```
Not necessarily two spaces, any strictly positive number is fine.

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+associativity of natural numbers under the addition. This time we will build
+the code line by line. You can load the file in each step.
+
+Enunciation of associativity
```
Is 'enunciation' a commonly used word? It might not help for a beginner if they have to look up new vocabulary just to get through the hello world. Perhaps "statement" is a better word to use (as in the statement of a theorem vs its proof).

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+
+  module hello-world-proof where
+
+Loading the file with ``C-c C-l`` should work. This is fon now the empty
```
Typo: fon

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+  open import Agda.Builtin.Nat using (Nat ; _+_)
+
+Next, we import the propositional equality between two terms ``_≡_``. This will
+allow us to state that two terms compute to the same normal form.
```
You are confusion propositional and definitional equality here. Definitional equality states (to a first approximation) that the terms compute to the same normal form, but cannot be presented internally in Agda. Propositional equality (which you are using here) represents that Agda can *prove* that the two terms are equal. For example, `m + n` and `n + m` are not definitionally equal but they are propositionally equal.

pnlph replied:
Yes, thanks. I added this description. But if the result of evaluating the propositional equality between two terms is `refl`, it means that they are definitionally equal, is it so?

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+.. code-block:: agda
+
+  +assoc-enun : Set
+  +assoc-enun = ∀ (x y z : Nat) → x + (y + z) ≡ (x + y) + z
```
It's not very common to explicitly define the statement of a proposition as an element of type `Set`, so I don't think it should be included here.

pnlph replied:
You are right; it is not very common. But I still think it is good to separate the statement lemma from its proof. Sometimes, the formalization of the statement of a theorem is hard enough and itself a result. This gives us also the opportunity of remarking in which universe the statement lives. Also that finding its proof is just finding an element of that type.

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+
+Now we have again one hole to resolve. If you load the file again, you will
+get the type of the term that should be in the hole
+``?0 : suc x + (y + z) ≡ suc x + y + z``. It seems like proving ``+assoc``
```
This is the first time that you rely essentially on type-level evaluation of a function, which is something that can be very non-obvious to newcomers. However, you ignore it here completely. It might make sense to introduce the normalize command `C-c C-n` or even the (ugly) prefix `C-u C-u` so you can see the computation happening: `suc x + (y + z)` computes to `suc (x + (y + z))` and `suc x + y + z` computes to `suc (x + y + z)`.

jespercockx replied:
It might also make sense to note that Agda prints the term `(x + y) + z` as `x + y + z` thanks to the infix statement in the builtin `Nat` module.

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+and load the file. Now the goal is just proving
+``?0 : x + (y + z) ≡ x + y + z``, which is the proof of ``+assoc x y z``.
+
+As it is structurally smaller than ``+assoc (suc x) y z``, we can recursively
```
Note that Agda does termination checking, and refer to the page on that feature.

jespercockx commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,417 @@
+First, we create a new file named ``hello-world-prog.agda`` with Emacs or Atom
+in a folder that we will refer to as our top-level folder.
+
+.. hint:: Agda programs are structured in :ref:`modules <module-system>`. The
```
Move this hint to where the first `module` declaration is explained.

jespercockx reviewed (changes requested):
Thanks for your work in writing this. I made a first pass and have some suggestions for changes.

pnlph committed:
fix Nat

https://github.com/agda/agda/pull/4756#discussion_r441607511

pnlph committed:
clarify module header

https://github.com/agda/agda/pull/4756#discussion_r441608000

pnlph committed:
strictly positive number of spaces

https://github.com/agda/agda/pull/4756#discussion_r441610589

pnlph committed:
enunciation → statement

https://github.com/agda/agda/pull/4756#discussion_r441611509

pnlph committed:
fix typo

https://github.com/agda/agda/pull/4756#discussion_r441611988

pnlph committed:
fix explanation definitional vs propositional equality

https://github.com/agda/agda/pull/4756#discussion_r441613668

pnlph committed:
fix explain computation steps

https://github.com/agda/agda/pull/4756#discussion_r441620914
and
https://github.com/agda/agda/pull/4756#discussion_r441622297

pnlph committed:
add termination checking explanation

https://github.com/agda/agda/pull/4756#discussion_r441623066

pnlph committed:
move hint modules

https://github.com/agda/agda/pull/4756#discussion_r441625449

UlfNorell commented:
I just fixed #3619 (#4760), so the discussion on installing `ieee754` can be moved to a less prominent place.

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+and a compatible version of the `standard library
+<https://github.com/agda/agda-stdlib/blob/master/notes/installation-guide.md>`_.
+You will also need an editor with *interactive* capabilities, currently
+Emacs via the :ref:`Emacs mode <emacs-mode>`, Atom via the
```
... currently supported are: Emacs ..., Atom ... and VSCode ...

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
 
-* Agda programs are structured in :ref:`modules <module-system>`. The
+.. hint:: Agda programs are structured in :ref:`modules <module-system>`. The
   first module in each file is the *top-level* module whose name
```
"first" is a bit misleading since it suggests the first element in a sequence.  However, the top module is the root of a tree.

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+   ``zero`` and ``suc`` from the Agda Builtin Library.
+3. declaration of the datatype ``Vec``.
+4. declaration of the empty vector constructor ``[]``.
+5. declaration of the append constructor ``_∷_``
```
Commonly, "append" is used for sequence concatenation.  Usual name: "cons".

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+6. specification of the :ref:`precedence` for the append operation
+
+.. note:: Paste or type the code above in a new file with extension ``.agda``.
+  Load the file with ``C-c C-l``. This also saves the file. You should see
```
Load the file (in Emacs, ``C-c C-l``).

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+bound of the types involved). We are now entitled to write the signature of
+our datatype: ``Vec (A : Set) : ℕ → Set``.
+
+To declare the datytype, we place its signature between the keywords ``data``
```
Typo here (try using a spell checker!).

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+‘Hello world’ proof
+===================
+
+We will see in this section how to enunciate and prove with Agda the
```
"enunciate" is a very uncommon word.  Better: "state and prove"

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+
+  open import Agda.Builtin.Equality using (_≡_)
+
+Now we can enunciate the associativity: given three (possibly different)
```
"state associativity" (no "the")

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+Now we can enunciate the associativity: given three (possibly different)
+natural numbers, adding the first to the addition of the second and the third
+computes to the same value as adding the addition of the first and the second
+to the third. We will name this statement ``+assoc-enun``.
```
More idiomatic would be `+-Associativity`.  Note that the naming convention is to use lowercase for values and capital for types.

andreasabel commented on doc/user-manual/getting-started/hello-world.lagda.rst:
```diff
@@ -7,35 +7,453 @@
+
+This code is self-contained and has several declarations:
+
+1. imports of the ``ÌO``, ``⊤`` and ``String`` datatypes from the
```
"Standard library" usually refers to the, er, standard library at https://github.com/agda/agda-stdlib.

pnlph commented:
> I just fixed #3619 (#4760), so the discussion on installing `ieee754` can be moved to a less prominent place.

Should the warning be removed completely or just turned into plain text?

UlfNorell commented:
I would suggest moving the warning about `ieee` to `tools/compilers`. There's already a line about it there which could be elaborated a little bit.

pnlph committed:
rewrite vector, editors

pnlph committed:
add `lookup` function

pnlph committed:
fix top-level description

https://github.com/agda/agda/pull/4756#discussion_r443192426

pnlph committed:
typo and state

pnlph committed:
+assoc → +-Assoc

https://github.com/agda/agda/pull/4756#discussion_r443193483

pnlph committed:
fix imported library

pnlph committed:
add emacs

https://github.com/agda/agda/pull/4756#discussion_r443192818

pnlph committed:
change append for cons

pnlph committed:
move warning about ieee to tools/compilers

https://github.com/agda/agda/pull/4756#issuecomment-647349685

pnlph committed:
fix warning title underline too short

pnlph committed:
minor primFloatEquality → Float

https://github.com/agda/agda/pull/4756#commitcomment-40077741

pnlph committed:
change URL Visual Studio Marketplace

URL agda-mode package atom

andreasabel commented:
I canceled the travis-ci for the latest push.  I think travis should not run on every push (costs lot of resources).  It should run once the PR is again ready for merging.
Maybe changing to "Draft" stops the CI?

pnlph commented:
Maybe would be good to change also the [HACKING.md -> Testing with Travis CI and GitHub Actions](https://github.com/agda/agda/blob/master/HACKING.md#testing-with-travis-ci-and-github-actions) section when #4729 is fixed. Then it would be possible to explain there the ways travis can be skipped? Eg. if the commit message contains `skip ci` (or `ci skip`?)

jespercockx commented:
@pnlph Are you still planning to continue working on this? It would be good to have this in before the release of 2.6.2, especially in the light of #4250. If you don't have the time, I could also try to take over and merge this in.

jespercockx commented:
I'm continuing this at https://github.com/agda/agda/pull/5223 so I'm closing this now.
